### ISSUE TITLE
False-positive TS7022 on destructured array element used as a 2-D index

### ISSUE THREAD

**Issue by yuki2006** [microsoft/typescript-go#4673]
## Steps to reproduce

```ts
// repro.ts
export function repro(board: number[][], pos: number | undefined): void {
  if (!pos) return;
  for (const pattern of [{ target: [-1, -1], intermediate: [-1, 0] }]) {
    const targetR = pos + pattern.target[0];
    const targetC = pos + pattern.target[1];
    if (board[targetR][targetC] !== 0) {}
    const [ir, ic] = pattern.intermediate;
    const midC = ic;
    if (board[ir][midC] === 0) {
    }
  }
}
```

```sh
npx tsc --noEmit repro.ts
```

No tsconfig or flags needed.

## Behavior with `typescript@6.0`

No error (6.0.3; also clean on 5.9.3).

## Behavior with `tsgo`

```
repro.ts(7,16): error TS7022: 'ic' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer.
```

Reproduces on 7.0.2 and 7.1.0-dev.20260719.1 (macOS darwin-arm64). There is no actual circularity: `ic` comes from destructuring `pattern.intermediate` (inferred `number[]`).

One observation that may help: removing just `| undefined` from `pos` — even while keeping the `if (!pos) return;` guard — makes the error disappear, so the control-flow-narrowed union type of `pos` appears to be essential to the false cycle. Every other line is also load-bearing (e.g. inlining `midC`, dropping the first `board[...][...]` access, or `unknown[][]` instead of `number[][]` each make it disappear).

Possibly related (all closed, but this still reproduces on today's nightly): #987, #3667, #3691, #943.

**Comment by ahejlsberg** [comment]
This is caused by a difference in the ported code for `isMatchingConstructorReference`. It's an easy fix, I will put up a PR.
